File: week_2/캐시.py
CACHE_HIT = 1
CACHE_MISS = 5

def solution(cacheSize, cities):
    if not cacheSize:
        return len(cities) * CACHE_MISS
    
    result = 0
    cache = dict()
    for at, city in enumerate(cities):
        city = city.lower()
        
        result += CACHE_HIT if city in cache else CACHE_MISS
        cache[city] = at
        
        if len(cache) > cacheSize:
            target = min(cache.items(), key = lambda x: x[1])[0]
            cache.pop(target)
    
    return result

File: week_2/캐시.js
const CACHE_HIT = 1;
const CACHE_MISS = 5;

function solution(cacheSize, cities) {
	if (!cacheSize) return cities.length * CACHE_MISS;

	const cache = {};
	const answer = cities.reduce((result, city, at) => {
		city = city.toLowerCase();

		const time = city in cache ? CACHE_HIT : CACHE_MISS;
		cache[city] = at;

		const entries = Object.entries(cache);
		if (entries.length > cacheSize) {
			const [target] = entries.reduce((min, cur) =>
				min[1] < cur[1] ? min : cur
			);
			delete cache[target];
		}

		return result + time;
	}, 0);

	return answer;
}
